### DISCUSSION TITLE
Transform html after next build without causing hydration error?

### DISCUSSION THREAD

**Discussion by michael-mcgreal** [vercel/next.js#45064] [Help] [answered]
I am attempting to run a cli command that **modifies HTML** (it's a [code syntax highlighter](https://torchlight.dev/docs/clients/cli)).

The cli program requires **HTML file input**, so I provide it the pathway to the **next build output** files `.next/server/pages`

Then modify the package.json built script to execute the command _after_ next build:
```
"scripts": {
    "build": "next build && yarn torchlight",
  },
```

When inspecting files in `.next/server/pages`, I can see that it transforms the HTML as desired (in place). 

However, it does not modify the associated .js files produced by nextjs (i.e. the files in `.next/static/chunks/static`)

Therefore, when viewing the app in the browser, the first frame renders the modified HTML, but when nextjs tries to hydrate the HTML using the .js file, it injects the non-modified html, and throws a hydration mismatch error.

To emphasize: if I disable javascript, I see the modified html that I desire. But enabling javascript will overwrite that html with the original output of next build.

Is there a different way to achieve this? Maybe by hooking into the build process, not after?

___

I am currently exploring [swc plugins](https://nextjs.org/docs/advanced-features/compiler#swc-plugins-experimental), similar to how 'modularize imports' transforms import statements before generating static html, but am hoping for a simpler simpler

**Comment by icyJoseph** [comment, marked as answer]
Why not do the syntax highlighting as part of the React application, and output highlighted HTML already? For instance, https://github.com/react-syntax-highlighter/react-syntax-highlighter? Or is it that you have content rendered through `dangerouslySetInnerHTML`? 

I think you can look into using `[supressHydrationError](https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning)`, but it has to be at node level IIRC.

**Reply by michael-mcgreal** [reply]
Hey again @icyJoseph! I see you everywhere – thanks for helping the community! I re-read your answers regularly.

I like how the highlighter you linked does _not_ 'dangerouslySetInnerHTML' – this is a great fallback.
I was wondering if suppressing the hydration error was possible – thanks for this reference! 

The reason for the "torchlight" syntax highlighter specifically is that it has delightful features that I couldn't find in other highlighters. Unfortunately, it does not maintain a js-client yet. (the closest workaround seems to be its remark plugin, although it has its own [module definition problems](https://github.com/torchlight-api/remark-torchlight/issues/1)).

If I am going to continue exploring this path of using the CLI package, I think my question can be refined to: "_Next build: How to run a CLI command after react-dom/server emits html and _before_ nextjs generates the associated javascript "instructions" files_". Though, I do not know if Next allows hooking into their build system like this.

Thanks again Joseph. Your knowledge of the internals is inspiring.

**Reply by michael-mcgreal** [reply]
Update: I was able to use the remark-torchlight plugin instead of the torchlight node cli (hacky workarounds). 

It feels strange to add .mdx only for its remark tooling that lets me access the html before react-dom/server gets it.

___

markdown file
```js
console.log('this is my code snippet')
console.log('written in mdx')
```
remark function inside getStaticProps
```
remark()
 .use(html) // convert it to html
 .use(torchlight) // syntax highlight – after html – before jsx
```
Then consume the output in a jsx component via 
```
<div dangerouslySetInnerHtml={{ __html: remarkOutput }} />
```

___

Thanks again joseph, your team is lucky to have you!
